### ISSUE TITLE
feat: replace HMR ids with MA ids in Human-GEM

### ISSUE THREAD

**Issue by haowang-bioinfo** [SysBioChalmers/Human-GEM#265]
### Description of the issue:
* This issue is a follow-up of #174 and aims to have a smooth transition from HMR ids to MA ids in `Human-GEM`

#### tentative implementation plan:
0. change prefix of MA reaction and metabolite ids to `MAR` and `MAM`, respectively
1. use MA reaction and metabolite ids formulated in #174 to replace `rxns` and `mets` fields in the model, respectively
2. swap the content between columns `rxns` and `rxnMAID` in `reactions.tsv`, then  rename column `rxnMAID` to `rxnRetired`
3. swap the content between columns `mets` and `metMAID` in `metabolites.tsv`, and rename column `metMAID` to `metRetired`
4. regenerate column `metNoComp` from new `mets` column

**I hereby confirm that I have:**
- [X] Done this analysis in the `master` branch of the repository
- [X] Checked that a similar issue does not exist already

**Comment by mihai-sysbio** [comment]
After some recent feedback from MetaNetX on the naming of the IDs, it has become clear that these should not contain any special characters. Therefore, the target prefixes are `MAR` and `MAM`.

**Comment by Hao-Chalmers** [comment]
> After some recent feedback from MetaNetX on the naming of the IDs, it has become clear that these should not contain any special characters. Therefore, the target prefixes are MAR and MAM.

Sounds good to me.

**Comment by JonathanRob** [comment]
Something to keep in mind, which is what @mihai-sysbio mentioned earlier: we need some system to keep track of which IDs have been used, so that retired IDs are not accidentally re-used later.

**Comment by mihai-sysbio** [comment]
I would guess that we could use the content of the annotation files for old IDs that have been used and removed, together with expanding the check that all the current IDs should exist in the annotation files, and none of these IDs should exist in the "used" annotation files.

**Comment by Hao-Chalmers** [comment]
Where to store these "used but removed" ids? @JonathanRob @mihai-sysbio

**Comment by mihai-sysbio** [comment]
> Where to store these "used but removed" ids? @JonathanRob @mihai-sysbio

How about `/data`?

**Comment by Hao-Chalmers** [comment]
> How about /data?

Do you mean an additional file?

**Comment by mihai-sysbio** [comment]
> Do you mean having an additional file? what is the spec (filename, file type)?

I'm thinking of files mirroring the contents of the annotation files `metabolites.tsv` and `reactions.tsv`, as these would capture both the "old" id and its mapping to other databases.

**Comment by Hao-Chalmers** [comment]
> I'm thinking of files mirroring the contents of the annotation files metabolites.tsv and reactions.tsv, as these would capture both the "old" id and its mapping to other databases.

This is relevant to the number 6 and 7 of the [10 Lessons](https://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.2001414) and should be considered further. 

But the others appear to be well-handled by this plan and [Met Atlas](https://metabolicatlas.org/).

- [x] Lesson 1. Credit any derived content using its original identifier
- [x] Lesson 2. Help local IDs travel well: Document prefix and patterns
- [x] Lesson 3. Opt for simple, durable web resolution
- [x] Lesson 4. Avoid embedding (too much) meaning or relying on it for uniqueness
- [x] Lesson 5. Design new identifiers for diverse uses by others
- [X] Lesson 8. Make URIs clear and findable
- [x] Lesson 9. Document the identifiers you issue and use
- [X] Lesson 10. Reference and display responsibly

**Comment by mihai-sysbio** [comment]
Here is how the plan looks like in my head: The files `expiredMetabolites.tsv` and `expiredReactions.tsv` will be placed in the `/data` folder. These files will have identical columns with the corresponding files in `/model`. To manually "expire" an MA identifier, after removing that component from the `yml` file, simply copy the row from the annotation file to the adequate `expired*.tsv`. To maintain consistency, expand the sanity check (all ids in the `yml` are in the `tsv`) to also check that none of `ids` in the model are in the `expired*.tsv`.

**Comment by mihai-sysbio** [comment]
As these `MA` ids are (re)used in other GEMs, things can get more complicated: each model might trigger the need for new IDs, and each model might "expire" IDs. So it looks to me like all the IDs need to be pooled together in a central place, most likely model-independent.

**Comment by Hao-Chalmers** [comment]
> So it looks to me like all the IDs need to be pooled together in a central place, most likely model-independent.

Fully agree, probably a repo under Met Atlas?

**Comment by mihai-sysbio** [comment]
> Fully agree, probably a repo under Met Atlas?

That's what make intuitive sense to me at the moment. My main concern is on maintaining the user-friendliness when eg wanting to coin a new MA id.

**Comment by Hao-Chalmers** [comment]
A tentative mechanism to keep track of IDs that have been used so that new IDs can be added uniquely and sequentially, while deprecated IDs are not accidentally re-used:
- have two new files `deprecatedMetabolites.tsv ` and `deprecatedReactions.tsv ` under `/data`
- keep `deprecatedMetabolites.tsv ` in the same column structure as `metabolites.tsv`, and `deprecatedReactions.tsv ` as `reactions.tsv`
- move rows from `deprecated***.tsv` to `***.tsv` when deprecating ids from the model
- add sanity checks ensuring model IDs should Not present in `deprecated***.tsv` files
- new MA IDs are formed by incrementing the highest number in the numeric part of existing IDs

**Comment by JonathanRob** [comment]
@Hao-Chalmers is your current suggestion is to keep this information on the Human-GEM repo?

Since MA IDs span other organisms beyond human, it seems to me that the maintenance of such IDs should not fall within the repository of a single species.

I liked your suggestion above to store this information somewhere within the Metabolic Atlas GitHub organization - they are **MA** IDs, after all. I'm not sure specifically where and how this information should be stored there, but I also agree with @mihai-sysbio that it should be implemented in a way that is user-friendly/intuitive.

**Comment by mihai-sysbio** [comment]
If all the MA ids are stored centrally, a script would probably have to fetch the latest version of this list before running the sanity check.

**Comment by mihai-sysbio** [comment]
I have created https://github.com/MetabolicAtlas/identifiers - let's try this out as a centralized ID maintenance repository.

**Comment by Hao-Chalmers** [comment]
> @Hao-Chalmers is your current suggestion is to keep this information on the Human-GEM repo?

@JonathanRob yes, I suggest to begin the implementation within Human-GEM repo.

> Since MA IDs span other organisms beyond human, it seems to me that the maintenance of such IDs should not fall within the repository of a single species.

Although the IDs span over multiple organisms and repos, forthcoming curations will be mainly within Human-GEM (at least during 4-6 months). With the `tsv` annotation files and manipulating functions in position and ready to be used, as well as the GH action checks, an implementation within Human-GEM would be straightforward, serving as an intermediate step toward a centralized workflow in the end.

> I liked your suggestion above to store this information somewhere within the Metabolic Atlas GitHub organization - they are MA IDs, after all. I'm not sure specifically where and how this information should be stored there, but I also agree with @mihai-sysbio that it should be implemented in a way that is user-friendly/intuitive.

I like the idea of having a centralized mechanism too, it's a good intension. But good intensions often delay things (they are too good to be actualized). It seems we have no ideas how to run it yet, except an empty repo.

If we begin the implementation within Human-GEM (as a test case and easy start), the required components (files, scripts) for creating a centralized pipeline surrounding the above [repo](https://github.com/MetabolicAtlas/identifiers) would be figured out and prepared in parallel (some may be directly copied from Human-GEM), and then hopefully it could be built in a few months.

**Comment by mihai-sysbio** [comment]
The issue as originally described has been resolved. However, a new problem arose regarding sanity checking of deprecated IDs that would be stored in a central repository. My proposal is to close this issue and open a new one for the new task.

**Comment by mihai-sysbio** [comment]
> The issue as originally described has been resolved. However, a new problem arose regarding sanity checking of deprecated IDs that would be stored in a central repository. My proposal is to close this issue and open a new one for the new task.

#407 was created to resolve the new spin-off issue, so as to be able to close the current one.
